### PR TITLE
add validator server for mixer

### PR DESCRIPTION
**Pull request by jmuk**
This PR adds the validator server as k8s external admission webhook
for mixer. Currently this will validate structurally -- i.e.
check if it can be converted into the proto.Message or not.

Further referential and semantic validation will come as another
PR.

The admission server logic comes from pilot/platform/kube/admit,
but had some tweaks to dropping the dependencies on pilot and/or
fitting with mixer's data structures. Ideally these implementations
should be unified into a single one, but for now, this would be
the reasonable solution.

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by jmuk** [comment]
/assign @mandarjog 
Please take a look -- this is mostly same as my previous PR in mixer repository.

File: mixer/cmd/server/cmd/validator.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+			runValidator(vc, kinds, printf, fatalf)
+		},
+	}
+	validatorCmd.PersistentFlags().StringVar(&vc.ServiceNamespace, "namespace", "istio-system", "the namespace where this webhook is deployed")
+	validatorCmd.PersistentFlags().StringVar(&vc.ServiceName, "webhook-name", "istio-mixer-webhook", "the name of the webhook")
+	validatorCmd.PersistentFlags().StringArrayVar(&vc.ValidateNamespaces, "target-namespaces", []string{},
+		"the list of namespaces where changes should be validated. Empty means to validate everything.")
+	validatorCmd.PersistentFlags().IntVarP(&vc.Port, "port", "p", 9099, "the port number of the webhook")
+	validatorCmd.PersistentFlags().StringVar(&vc.SecretName, "secret-name", "", "The name of k8s secret where the certificates are stored")
+	validatorCmd.PersistentFlags().DurationVar(&vc.RegistrationDelay, "registration-delay", 5*time.Second, "Time to delay webhook registration after starting webhook server")
```
Comment (mandarjog) [inline mixer/cmd/server/cmd/validator.go]: why is this needed ?

File: mixer/cmd/server/cmd/validator.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	return validatorCmd
+}
+
+func createK8sClient() (*kubernetes.Clientset, error) {
+	// Validator needs to run within a cluster. It should work as long as InClusterConfig is valid.
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(config)
```
Comment (mandarjog) [inline mixer/cmd/server/cmd/validator.go]: just 
`return kubernetes.NewForConfig(config)` ?

File: mixer/cmd/server/cmd/validator.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func runValidator(vc crd.ControllerOptions, kinds map[string]proto.Message, printf, fatalf shared.FormatFn) {
+	client, err := createK8sClient()
+	if err != nil {
+		printf("Failed to create kubernetes client: %v", err)
+		printf("Starting plain http server, but external admission hook is not enabled")
+		client = nil
+	}
+	vs, err := crd.NewController(client, vc)
```
Comment (mandarjog) [inline mixer/cmd/server/cmd/validator.go]: Can you elaborate on how it is ok for `client` to be `nil` ?

File: mixer/pkg/config/crd/admit.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+	}
+	tlsConfig, err := makeTLSConfig(serverCert, serverKey, apiServerCACert)
+	if err != nil {
+		return nil, nil, err
+	}
+	return tlsConfig, caCert, nil
+}
+
+// Run implements the admission controller run loop.
+func (ac *AdmissionController) Run(stop <-chan struct{}) {
```
Comment (mandarjog) [inline mixer/pkg/config/crd/admit.go]: The referenced issue has been fixed. https://github.com/kubernetes/kubernetes/pull/50476
Either fix it now, or file an issue jointly with pilot to address it afterwards.

File: mixer/pkg/config/crd/admit.go [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+	ServiceName string
+
+	// ServiceNamespace is the namespace of the webhook service.
+	ServiceNamespace string
+
+	// ValidateNamespaces is a list of names to validate. Any
+	// namespace not in this list is unconditionally validated as
+	// good. This is useful when multiple validators are running in
+	// the same cluster managing different sets of namespaces
+	// (e.g. shared test clusters).
```
Comment (mandarjog) [inline mixer/pkg/config/crd/admit.go]: This option is very test-infra specific. If it is not meant for prod use, we should make it clear in the comments.

File: mixer/pkg/config/crd/admit.go [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+type AdmissionController struct {
+	client  kubernetes.Interface
+	options ControllerOptions
+}
+
+// GetAPIServerExtensionCACert gets the Kubernetes aggregate apiserver
+// client CA cert used by the "GenericAdmissionWebhook" plugin
+// admission controller.
+//
+// NOTE: this certificate is provided kubernetes. We do not control
```
Comment (mandarjog) [inline mixer/pkg/config/crd/admit.go]: nit: provided `to` K8s or provided `by` k8s?

File: mixer/pkg/config/crd/admit.go [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+	<-stop
+	server.Close() // nolint: errcheck
+}
+
+// Unregister unregisters the external admission webhook
+func (ac *AdmissionController) unregister(client admissionClient.ExternalAdmissionHookConfigurationInterface) error {
+	return client.Delete(ac.options.ExternalAdmissionWebhookName, nil)
+}
+
+// Register registers the external admission webhook for pilot
```
Comment (mandarjog) [inline mixer/pkg/config/crd/admit.go]: pilot -> mixer (or istio)

File: mixer/pkg/config/crd/admit.go [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+
+// ServeHTTP implements the external admission webhook.
+func (ac *AdmissionController) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	glog.V(4).Infof("AdmissionController ServeHTTP request=%#v", r)
+
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
```
Comment (mandarjog) [inline mixer/pkg/config/crd/admit.go]: log error on unable to read?

File: mixer/cmd/server/cmd/validator.go (hunk 1 above)
Comment (jmuk) [reply mixer/cmd/server/cmd/validator.go]: This is the delay before registering itself as a webhook, without the delay, the registration could happen too early for k8s apiserver to fail to find where the webhook is (i.e. `Service` existence).
This is the same one as pilot webhook has, see https://github.com/istio/istio/blob/master/pilot/cmd/pilot-discovery/main.go#L278

File: mixer/cmd/server/cmd/validator.go (hunk 2 above)
Comment (jmuk) [reply mixer/cmd/server/cmd/validator.go]: ah, right. done

File: mixer/cmd/server/cmd/validator.go (hunk 3 above)
Comment (jmuk) [reply mixer/cmd/server/cmd/validator.go]: oops, it's not okay now (it was okay with my previous implementation...), fixed.

File: mixer/pkg/config/crd/admit.go (hunk 5 above)
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: added here, and the flag description.

File: mixer/pkg/config/crd/admit.go (hunk 6 above)
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: by. added.

File: mixer/pkg/config/crd/admit.go (hunk 4 above)
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: It's not good to fix here locally, better to be aligned with pilot webhooks. Filed https://github.com/istio/istio/issues/1795, and change the TODO reference to the new issue for the tracking. I'll keep my eyes on it.

File: mixer/pkg/config/crd/admit.go (hunk 7 above)
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: done

File: mixer/pkg/config/crd/admit.go (hunk 8 above)
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: done

File: mixer/pkg/config/crd/admit.go [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crd
+
+// Note: this file is copied from pilot/platform/kube/admit. Ideally we should
+// make a common component for this server side part, accepting component-specific
+// validation logic.
+// TODO: do so.
```
[the remark this answers is not in the document]
Comment (jmuk) [reply mixer/pkg/config/crd/admit.go]: done

File: mixer/cmd/server/cmd/validator.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	validatorCmd.PersistentFlags().StringVar(&vc.ServiceName, "webhook-name", "istio-mixer-webhook", "the name of the webhook")
+	validatorCmd.PersistentFlags().StringArrayVar(&vc.ValidateNamespaces, "target-namespaces", []string{},
+		"the list of namespaces where changes should be validated. Empty means to validate everything. Used for test only.")
+	validatorCmd.PersistentFlags().IntVarP(&vc.Port, "port", "p", 9099, "the port number of the webhook")
+	validatorCmd.PersistentFlags().StringVar(&vc.SecretName, "secret-name", "", "The name of k8s secret where the certificates are stored")
+	validatorCmd.PersistentFlags().DurationVar(&vc.RegistrationDelay, "registration-delay", 5*time.Second, "Time to delay webhook registration after starting webhook server")
+	return validatorCmd
+}
+
+func createK8sClient() (*kubernetes.Clientset, error) {
```
[the remark this answers is not in the document]
Comment (jmuk) [reply mixer/cmd/server/cmd/validator.go]: hmm, I see. done.

**Comment by jmuk** [comment]
btw, I've noticed it's been missing the actual integration with the commandline -- now confirmed working.
